File: src/main/java/com/y271727uy/ranch_festival/dimension/DimensionRegistry.java
package com.y271727uy.ranch_festival.dimension;

import java.util.LinkedHashSet;
import java.util.Set;
import net.minecraft.resources.ResourceKey;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;

public final class DimensionRegistry {
    private static final Set<DimensionDefinition> DEFINITIONS = new LinkedHashSet<>();

    private DimensionRegistry() {
    }

    public static DimensionDefinition register(DimensionDefinition definition) {
        DEFINITIONS.add(definition);
        return definition;
    }

    public static DimensionDefinition findByDimension(ResourceKey<Level> dimension) {
        for (DimensionDefinition definition : DEFINITIONS) {
            if (definition != null && dimension != null && dimension.equals(definition.getStructureDimension())) {
                return definition;
            }
        }
        return null;
    }

    public static DimensionDefinition findByInvitation(ItemStack itemStack) {
        if (itemStack == null || itemStack.isEmpty()) {
            return null;
        }

        for (DimensionDefinition definition : DEFINITIONS) {
            if (definition != null && definition.matchesInvitation(itemStack.getItem())) {
                return definition;
            }
        }
        return null;
    }

    public static Set<DimensionDefinition> all() {
        return Set.copyOf(DEFINITIONS);
    }
}

File: src/main/kotlin/com/y271727uy/ranch_festival/dimension/DimensionRegistry.kt
package com.y271727uy.ranch_festival.dimension

import java.util.Collections
import java.util.LinkedHashSet
import net.minecraft.resources.ResourceKey
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level

object DimensionRegistry {
    private val definitions: MutableSet<DimensionDefinition> = LinkedHashSet()

    @JvmStatic
    fun register(definition: DimensionDefinition): DimensionDefinition {
        definitions.add(definition)
        return definition
    }

    @JvmStatic
    fun findByDimension(dimension: ResourceKey<Level>?): DimensionDefinition? {
        if (dimension == null) {
            return null
        }

        return definitions.firstOrNull { it.structureDimension == dimension }
    }

    @JvmStatic
    fun findByInvitation(itemStack: ItemStack?): DimensionDefinition? {
        if (itemStack == null || itemStack.isEmpty) {
            return null
        }

        return definitions.firstOrNull { it.matchesInvitation(itemStack.item) }
    }

    @JvmStatic
    @Suppress("unused")
    fun all(): Set<DimensionDefinition> {
        return Collections.unmodifiableSet(LinkedHashSet(definitions))
    }
}


